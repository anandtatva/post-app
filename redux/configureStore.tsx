import { combineReducers, createStore } from "redux";
import { persistStore, persistReducer } from "redux-persist";
import AsyncStorage from "@react-native-async-storage/async-storage";
import PostListSlice from "./reducer/post-list-slice";

const persistConfig = {
    key: "post-list",
    storage: AsyncStorage
}

const rootReducer = combineReducers({
    posts: PostListSlice
})

const persistReducers = persistReducer(persistConfig, rootReducer);

const store = createStore(persistReducers);

const persistor = persistStore(store);

export {
    store,
    persistor
}