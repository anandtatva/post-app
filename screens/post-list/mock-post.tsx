import { PostType } from "../../types/post-type";

export const mockPostList: PostType[] = [
    { title: "New Song1", url: "http://new-song1.com", created_at: "1", author: "Jack1" },
    { title: "New Song2", url: "http://new-song2.com", created_at: "2", author: "Jack2" }
]