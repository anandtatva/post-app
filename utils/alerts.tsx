import { Platform, Alert } from "react-native";

export const showAlert = (message: string) => {
    if (Platform.OS === "web") {
        alert(message)
    } else Alert.alert(message);

}
