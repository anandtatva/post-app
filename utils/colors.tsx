export const colors = {
    primary: "#000",
    secondary: "#FFF",
    gray: "grey",
}