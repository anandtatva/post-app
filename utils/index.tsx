export * from "./alerts";
export * from "./strings";
export * from "./commonStyle";
export * from "./colors";
export * from "./utils"