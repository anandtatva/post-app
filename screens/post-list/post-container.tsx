import { Text, TouchableOpacity, View, StyleSheet } from "react-native"
import { PostType } from "../../types/post-type";
import { FC } from "react";
import { colors } from "../../utils";


interface PostContainerProps {
    index: number;
    item: PostType;
    onNext: (item: PostType) => void;
}
const PostContainerView: FC<PostContainerProps> = ({
    item, index, onNext
}) => {
    const handleNext = () => { onNext(item) }
    return (
        <TouchableOpacity
            onPress={handleNext}
            testID={`Post-Row-${index}`}
            style={styles.conatiner}
        >
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Title`}>Title</Text>
                <Text style={styles.text} numberOfLines={1} testID={`Title-${index}`}>{item.title}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Author`}>Author</Text>
                <Text style={styles.text} numberOfLines={1} testID={`Author-${index}`}>{item.author}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`URL`}>URL</Text>
                <Text style={styles.text} numberOfLines={1} testID={`URL-${index}`}>{item.url} {item.url}{item.url}</Text>
            </View>
            <View style={styles.detailsContainer}>
                <Text style={styles.title} testID={`Created-At`}>Created At</Text>
                <Text style={styles.text} numberOfLines={1} testID={`Created-At-${index}`}>{item.created_at}</Text>
            </View>
        </TouchableOpacity>
    )
}

export default PostContainerView;

const styles = StyleSheet.create({
    detailsContainer: {
        paddingVertical: 2,
        flexDirection: "row",
        alignItems: "center",
        justifyContent: "space-between",
    },
    conatiner: {
        flex: 1,
        backgroundColor: colors.primary,
        paddingHorizontal: 10,
        paddingVertical: 10
    },
    title: {
        flex: 2,
        fontWeight: "bold",
        color: colors.secondary

    },
    text: {
        flex: 4,
        paddingLeft: 5,
        color: colors.secondary,
        fontWeight: "300",
    }
})