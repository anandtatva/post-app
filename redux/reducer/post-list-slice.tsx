import { createSlice } from "@reduxjs/toolkit";
import { PostType } from "../../types/post-type";
import { mockPostList } from "../../screens/post-list/mock-post";
interface PostListStateProps {
    totalPage: number;
    currentPage: number;
    postList: PostType[];
}

const initialState: PostListStateProps = {
    postList: [],
    currentPage: 0,
    totalPage: 0

}
const PostListSlice = createSlice({
    name: "post-list-slice",
    initialState,
    reducers: {
        setCurrentPage: (state, action) => {
            state.currentPage = action.payload
        },
        setTotalPage: (state, action) => {
            state.totalPage = action.payload
        },
        setPostListPage: (state, action) => {
            state.postList = action.payload
        },
        setAllData: (state, action) => {
            state.postList = action.payload.currentPage === 0 ? action.payload.postList : [...state.postList, ...action.payload.postList]
            state.totalPage = action.payload.totalPage
            state.currentPage = action.payload.currentPage
        }
    }
})

export const { setAllData, setCurrentPage, setTotalPage, setPostListPage } = PostListSlice.actions;

export default PostListSlice.reducer;