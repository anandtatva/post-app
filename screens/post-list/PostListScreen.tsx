import { Text, View, FlatList, StyleSheet } from "react-native"
import { mockPostList } from "./mock-post";
import PostContainerView from "./post-container";
import { PostType } from "../../types/post-type";
import { useSelector, useDispatch } from "react-redux";
import { colors, strings, commonStyle, showAlert } from "../../utils";
import { getPostAPIs } from "../../services/api";
import { setTotalPage, setAllData } from "../../redux/reducer/post-list-slice";
import { useEffect } from "react";

const PostListScreen = ({ navigation }) => {
    const postList = useSelector(state => state.posts)
    const dispatch = useDispatch();
    useEffect(() => {
        _loadPostList(0);
    }, [])
    const _loadPostList = async (page: number) => {
        const resp = await getPostAPIs(page);
        if (resp.status) {
            const { data, currentPage, totalPage } = resp;
            dispatch(setAllData({
                totalPage,
                postList: data,
                currentPage
            }))
            if (totalPage > currentPage + 1) {
                setTimeout(() => {
                    _loadPostList(page + 1)
                }, 10000);
            }
        } else showAlert(resp.message)
    }
    const renderItem = ({ item, index }: { item: PostType; index: number; }) => (
        <PostContainerView
            onNext={(item: PostType) => { navigation.navigate("PostDetailsScreen", { item }) }}
            item={item}
            index={index}
        />
    )
    return (
        <FlatList
            ListHeaderComponent={() => <View style={{
                paddingVertical: 10,
            }}>
                <Text style={styles.emptyText} testID="Current-Total-Post">Current Total Post:- {postList.postList.length}</Text>
                <Text style={styles.emptyText} testID="Current-Page-Total-Page">Current Page:- {postList?.currentPage + 1} Total Page:-  {postList?.totalPage}</Text>
            </View>}
            ItemSeparatorComponent={() => <View style={styles.separator} />}
            data={postList?.postList || []}
            keyExtractor={(_, index) => String(index)}
            renderItem={renderItem}
            ListEmptyComponent={() => (
                <Text style={styles.emptyText} testID="Empty-Post-List" >{strings.emptyPostList}</Text>
            )}
        />
    )
}

export default PostListScreen;

const styles = StyleSheet.create({
    separator: {
        borderBottomWidth: 1.5,
        borderColor: colors.gray
    },
    emptyText: {
        fontWeight: "bold",
        textAlign: "center"
    }
})