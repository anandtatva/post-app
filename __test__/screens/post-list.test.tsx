import { cleanup, fireEvent, render, waitFor } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { store } from "../../redux/configureStore";
import PostListScreen from "../../screens/post-list/PostListScreen";
import { setAllData } from "../../redux/reducer/post-list-slice";
import { strings } from "../../utils";
import { mockPostList } from "../../screens/post-list/mock-post";


jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native")
    return {
        ...actualNav,
        useNavigation: () => ({
            navigate: jest.fn()
        }),
        useFocusEffect: () => jest.fn(),
        useRoute: () => ({
            params: {}
        })
    }

})
describe("Screen: Post List", () => {
    beforeEach(() => {
        store.dispatch(setAllData({
            currentPage: 0,
            totalPage: 0,
            postList: []
        }))
    })
    afterEach(() => {
        cleanup()
        jest.clearAllMocks()
    })
    it("renders correctly", () => {
        const navigation = { navigate: jest.fn() }
        const tree = render(
            <Provider store={store}>
                <PostListScreen navigation={navigation} />
            </Provider>
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    it("show empty message if no post found", () => {
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <PostListScreen navigation={navigation} />
            </Provider>
        )
        const emptyText = getByTestId("Empty-Post-List");
        expect(emptyText.children[0]).toBe(strings.emptyPostList)
    })
    it("move to details page on press post row", () => {
        store.dispatch(setAllData({
            currentPage: 1,
            totalPage: 1,
            postList: mockPostList
        }))
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <PostListScreen navigation={navigation} />
            </Provider>
        )
        const postRow1 = getByTestId("Post-Row-1");
        fireEvent.press(postRow1)
        expect(navigation.navigate).toHaveBeenCalledWith("PostDetailsScreen", { item: mockPostList[1] })
    })
    it("page should be show correct total page, current page and total post", () => {
        const data = {
            currentPage: 0,
            totalPage: 1,
            postList: mockPostList
        }
        store.dispatch(setAllData(data))
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <PostListScreen navigation={navigation} />
            </Provider>
        )
        const currentTotalPost = getByTestId("Current-Total-Post");
        const currentPageNTotalPage = getByTestId("Current-Page-Total-Page");

        expect(currentTotalPost.children[1]).toBe(String(data.postList.length)) // current total post
        expect(currentPageNTotalPage.children[1]).toBe(String(data.currentPage + 1)) // total page
        expect(currentPageNTotalPage.children[3]).toBe(String(data.totalPage)) // current page
    })
})