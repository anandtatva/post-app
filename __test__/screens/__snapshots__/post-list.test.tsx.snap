// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen: Post List renders correctly 1`] = `
<RCTScrollView
  ItemSeparatorComponent={[Function]}
  ListEmptyComponent={[Function]}
  ListHeaderComponent={[Function]}
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
          testID="Current-Total-Post"
        >
          Current Total Post:- 
          0
        </Text>
        <Text
          style={
            Object {
              "fontWeight": "bold",
              "textAlign": "center",
            }
          }
          testID="Current-Page-Total-Page"
        >
          Current Page:- 
          1
           Total Page:-  
          0
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
      testID="Empty-Post-List"
    >
      No Post Found
    </Text>
  </View>
</RCTScrollView>
`;
